### ISSUE TITLE
Enabling the autoscaler feature restarts credhub and uaa

### ISSUE THREAD

**Issue by jandubois** [cloudfoundry-incubator/kubecf#1487]
**Describe the bug**

I installed kubecf with `make kubecf-apply` (uses Diego). Pushed and tested 12factor sample app.

Then upgraded via `FEATURE_AUTOSCALER=1 make kubecf-apply`. This first restarted both `bosh-dns` pods, which may be necessary (I don't know). But then it also stopped `credhub` and `uaa`, which should not be affected.

Eventually `credhub` and `uaa` come back, and all autoscaler pods start, and everything seems fine. The 12factor app remained accessible during the whole time.

**Expected behavior**

I expect `credhub` and `uaa` not to be restarted when the only requested change is enabling autoscaler.

**Comment by viovanov** [comment]
@jandubois it's likely that there's a new client added to UAA to facilitate authentication across components, which would mean this behavior is expected.

**Comment by andreas-kupries** [comment]
# TLDR

Not a bug, IMHO. Closing this ticket.

# Deeper notes

Looking at the statefulsets for `uaa` and `credhub`, before and after the upgrade which activated auto-scaler, I note that these sts got bumped from `v1` to `v2` as part of the upgrade. Their references to the `ig-resolved.XXX-vN` secrets changed accordingly, also.

This version bump is then responsible for the pods to rotate.

Now, why is there a version bump ?

Looking at the referenced `ig-resolved.*-v*` secrets for these statefulsets and versions I find actual and sensible differences in them.

First `uaa`:
```
         clients:
+          app_autoscaler:
+            authorities: cloud_controller.read,cloud_controller.admin,uaa.resource
+            authorized-grant-types: client_credentials
+            scope: cloud_controller.read,cloud_controller.write,openid,scim.read
+            secret: iPfhsS0tshF4WBPkCb1N95EPdyGW0CkNQUjk4Swi5b9nUZdEqRwzVZKlIkArpZL1
```

This is actually as @viovanov suspected, a month ago. I now confirm.

And `credhub`:
 ```
       quarks:
         consumes:
           postgres:
-            address: ""
-            instances: null
-            properties: null
+            address: asdatabase
+            instances:
+            - address: asdatabase-0
+              az: ""
+              bootstrap: true
+              id: asdatabase-0
+              index: 0
+              instance: 0
+              name: asdatabase-postgres
+              networks: null
+            properties:
+              databases:
+                databases:
+                - name: autoscaler
+                  tag: default
+                port: 5432
+                roles:
+                - name: postgres
+                  password: g7rpEjcaAr9CWolcw7chSCROBucWEqFUwHHQh9GcfJAY3yE2WCIcBDgEa1elppHC
+                  tag: default
+                tls:
+                  ca: &fdf73d2ee2591a59140f0a6f5d24e18ba24487ec |
+                    -----BEGIN CERTIFICATE-----
+                    MIIC+jCCAeKgAwIBAgIUPEBo/yKCuBEr3qcgdM49yhRijZowDQYJKoZIhvcNAQEL
[...]
 ```

So, activating the autoscaler does reconfigure these statefulsets, adding in autoscaler-specific information.
Having these pods rotate is therefore correct.

__Not a bug__ :tada:  Closing.

**Comment by mook-as** [comment]
Wait, why does enabling autoscaler make credhub use _its_ postgres database? Does that mean existing credhub data is ignored?

**Comment by andreas-kupries** [comment]
Honestly, I have no idea. That might be worth a separate investigation. :bell: @viovanov @satadruroy @gaktive

**Comment by jandubois** [comment]
> why does enabling autoscaler make credhub use _its_ postgres database?

That doesn't make any sense, and I can't see any code in the template that would do that. @andreas-kupries are you sure that snippet is from the resolved credhub secret?

**Comment by andreas-kupries** [comment]
The diff was from
```
--- C-0001/SECRETS/ig-resolved.credhub-v1.txt   2020-12-04 11:07:40.749175571 -0800
+++ C-0003/SECRETS/ig-resolved.credhub-v2.txt   2020-12-04 11:08:01.513426387 -0800
```

I can redo this when I am done with re-testing #1594 .
See also the script I posted to cap-devel, on how I am pulling the information.

**Comment by andreas-kupries** [comment]
The v1 was from before the upgrade, and v2 from after.

**Comment by andreas-kupries** [comment]
That said, I am __in full support__ of somebody else falsifying this as well in their setup.

**Comment by andreas-kupries** [comment]
Going back to this. I will now start with auto-scaler from scratch, and then look at the uaa and credhub ig-resolved.
Then drop that deployment and re-deploy without auto-scaler, and again look at the ig-resolved.
I.e. instead of upgrade 2 more independent (*) deployments.

(*) I will keep the minikube and cf-operator the same. Save me some time in copying images around internally.

**Comment by andreas-kupries** [comment]
As part of the above ticket re-opened and back to in progress.
Note: I am working with master commit 024937cc3094d86d9d9b7748ce629b31ac4a3cbc.

**Comment by mook-as** [comment]
It's probably just BOSH links in action:
﻿https://github.com/pivotal/credhub-release/blob/1578b01af052da1d3a894b4c9ed5d9f444d0a9ab/jobs/credhub/spec#L79-L82
```yaml
consumes:
- name: postgres
  type: database
  optional: true
```

That… doesn't resolve question about the data, though.

**Comment by mook-as** [comment]
Looking at that release some more,
- The postgres usage is probably [opt-in](https://github.com/pivotal/credhub-release/blob/1578b01af052da1d3a894b4c9ed5d9f444d0a9ab/jobs/credhub/templates/application_spring.yml.erb#L51).
- cf-deployment uses [mysql by default](https://github.com/cloudfoundry/cf-deployment/blob/b119f228bfb28c8e8d5514a5ed46f1f5b4494510/cf-deployment.yml#L1790).

So we're probably fine?

**Comment by andreas-kupries** [comment]
I confirm the difference for credhub's ig-resolved using the separate deployments.
There are more differences now, however that is because the two deployments have separate and different certs.
Ditto confirm uaa.

So, pretty sure now that the difference is real. If that difference will make trouble with credhub, that I still do not know.

I will keep this open now so that we can talk Monday in the standup.

**Comment by andreas-kupries** [comment]
Created issue #1649 for the credhub/autoscaler semi-clash.
Closing this is as done (again).
